### PR TITLE
GH-39333: [C++] Don't use "if constexpr" in lambda

### LINKED ISSUE #39333: [CI][Python][Packaging] Failed to build wheel for Windows since ListView merge
### Describe the bug, including details regarding any error messages, version, and platform.

Failed log:

https://github.com/ursacomputing/crossbow/actions/runs/7272850266/job/19815700878#step:7:768

```text
C:/arrow/cpp/src/arrow/array/array_nested.cc(291): error C2065: 'validity': undeclared identifier (compiling source file C:\arrow-build\src\arrow\CMakeFiles\arrow_shared.dir\Unity\unity_0_cxx.cxx) [C:\arrow-build\src\arrow\arrow_shared.vcxproj]
  C:/arrow/cpp/src/arrow/array/array_nested.cc(660): note: see reference to function template instantiation 'arrow::Result<std::shared_ptr<arrow::Array>> arrow::`anonymous-namespace'::FlattenListViewArray<arrow::ListViewArray,false>(const ListViewArrayT &,arrow::MemoryPool *)' being compiled
          with
          [
              ListViewArrayT=arrow::ListViewArray
          ] (compiling source file C:\arrow-build\src\arrow\CMakeFiles\arrow_shared.dir\Unity\unity_0_cxx.cxx)
  memory_pool.cc
C:/arrow/cpp/src/arrow/array/array_nested.cc(291): error C2065: 'list_view_array_offset': undeclared identifier (compiling source file C:\arrow-build\src\arrow\CMakeFiles\arrow_shared.dir\Unity\unity_0_cxx.cxx) [C:\arrow-build\src\arrow\arrow_shared.vcxproj]
```

Related change: https://github.com/apache/arrow/pull/35345

### Component(s)

C++, Continuous Integration, Packaging, Python

### PR DESCRIPTION
**Pull request by kou**
### Rationale for this change

It seems that it's not portable. At least it doesn't work as expected with Visual Studio 2017:

    C:/arrow/cpp/src/arrow/array/array_nested.cc(291): error C2065: 'validity': undeclared identifier (compiling source file C:\arrow-build\src\arrow\CMakeFiles\arrow_shared.dir\Unity\unity_0_cxx.cxx) [C:\arrow-build\src\arrow\arrow_shared.vcxproj]
      C:/arrow/cpp/src/arrow/array/array_nested.cc(660): note: see reference to function template instantiation 'arrow::Result<std::shared_ptr<arrow::Array>> arrow::`anonymous-namespace'::FlattenListViewArray<arrow::ListViewArray,false>(const ListViewArrayT &,arrow::MemoryPool *)' being compiled
              with
              [
                  ListViewArrayT=arrow::ListViewArray
              ] (compiling source file C:\arrow-build\src\arrow\CMakeFiles\arrow_shared.dir\Unity\unity_0_cxx.cxx)
      memory_pool.cc
    C:/arrow/cpp/src/arrow/array/array_nested.cc(291): error C2065: 'list_view_array_offset': undeclared identifier (compiling source file C:\arrow-build\src\arrow\CMakeFiles\arrow_shared.dir\Unity\unity_0_cxx.cxx) [C:\arrow-build\src\arrow\arrow_shared.vcxproj]
### What changes are included in this PR?

Avoid "if constexpr" in lambda.

### Are these changes tested?

Yes.

### Are there any user-facing changes?

No.
* Closes: #39333

### CODE REVIEW

**Comment by kou** [comment]
@github-actions crossbow submit wheel-windows-*

**Comment by kou** [comment]
@github-actions crossbow submit wheel-windows-*

**Comment by kou** [comment]
@felipecrv This is happen after #35345 is merged.

File: cpp/src/arrow/array/array_nested.cc [hunk 1]
```diff
@@ -282,18 +281,24 @@ Result<std::shared_ptr<Array>> FlattenListViewArray(const ListViewArrayT& list_v
     }
   }
 
-  const auto* validity = list_view_array.data()->template GetValues<uint8_t>(0, 0);
   const auto* offsets = list_view_array.data()->template GetValues<offset_type>(1);
   const auto* sizes = list_view_array.data()->template GetValues<offset_type>(2);
 
-  auto is_null_or_empty = [&](int64_t i) {
-    if constexpr (HasNulls) {
+  std::function<bool(int64_t)> is_null_or_empty;
```
Comment (pitrou) [inline cpp/src/arrow/array/array_nested.cc:287]: This is complicated. Instead, how about just:
```c++
  auto is_null_or_empty = [&](int64_t i) {
    if (HasNulls && !bit_util::GetBit(validity, list_view_array_offset + i)) {
      return true;
    }
    return sizes[i] == 0;
  };
```

**Comment by pitrou** [comment]
@github-actions crossbow submit -g cpp wheel-windows-*

File: cpp/src/arrow/array/array_nested.cc (hunk 1 above)
Comment (felipecrv) [reply cpp/src/arrow/array/array_nested.cc]: +1.

The compiler will constant-fold it when the lambda gets inlined. Putting it in `std::function` would prevent inlining.

**Pull request merged**